tests/agents/test_acu_agent.py

```diff
@@ -72,3 +119,71 @@ def test_tracks():
     points = next(iter(g))
     drivers.get_track_points_text(points, text_block=True,
                                   timestamp_offset=3)
+
+
+def test_hwp_iface_ranges_math():
+    misc = [
+        (10, 20), (30, 40), (20, 25), (-10, 11)
+    ]
+    assert hwp_iface._simplify_ranges(*misc) == [(-10, 25), (30, 40)]
+    for a, b, result in [
+            ((0, 10), (11, 20), False),
+            ((0, 10), (1, 1), True),
+            ((0, 10), (2, 10), True),
+            ((0, 10), (-1, 4), False),
+            ((0, 10), (8, 11), False),
+    ]:
+        assert hwp_iface._range_contains(a, b) == result
+
+
+def _gen_rules(*args):
+    def listify(x):
+        return [x] if isinstance(x, str) else x
+    rules = []
+    for el0, el1, grip, spin, el, az, th in args:
+        rules.append({
+            'el_range': [el0, el1],
+            'grip_states': listify(grip),
+            'spin_states': listify(spin),
+            'allow_moves': {
+                'el': el,
+                'az': az,
+                'third': th,
+            }})
+    return {'tolerance': .1,
+            'rules': rules}
+
+
+def test_hwp_iface():
+    cfg = HWP_IFACE_TEST_CONFIG
+    mo_rules = hwp_iface.HWPInterlocks.from_dict(cfg)
+
+    def ex(d, item=0):
+        return tuple([d[k][item] for k in ['el', 'az', 'third']])
+
+    mo_rules = hwp_iface.HWPInterlocks.from_dict(_gen_rules(
+        (40, 70, 'gripped', 'not_spinning', True, False, False),
+        (50, 60, 'ungripped', 'not_spinning', False, True, True),
+        (55, 62, 'ungripped', 'spinning', True, True, False),
+    ))
```

Comment:
I'm not quite following this, why define `HWP_IFACE_TEST_CONFIG` and generate the `HWPInterlocks` from that config if we just redefine them on line 164 from this list?

Reply:
Added comments about this ... I think the idea was to test parsing a "full config" and then test some specific combination of tricky rules.

Reply:
Ah, that makes sense. I'd probably split it into two tests personally -- one that makes sure `mo_rules` has the form you expect when loading from `cfg`, and the other that tests the range combination rules. Your call though.